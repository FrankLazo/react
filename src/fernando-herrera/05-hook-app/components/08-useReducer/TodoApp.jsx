import { useEffect, useReducer } from 'react';
import TodoAdd from './TodoAdd';
import TodoList from './TodoList';
import todoReducer from './todoReducer';

const init = () => (
	JSON.parse(localStorage.getItem('todos')) || []
);

const TodoApp = () =>
{
	const [todos, dispatch] = useReducer(todoReducer, [], init);

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	const handleDelete = (todoId) =>
	{
		const action = {
			type:    'delete',
			payload: todoId,
		};

		dispatch(action);
	};

	const handleToggle = (todoId) =>
	{
		dispatch({
			type:    'toggle',
			payload: todoId,
		});
	};

	const handleAddTodo = (newTodo) =>
	{
		dispatch({
			type:    'add',
			payload: newTodo,
		});
	};

	return (
		<div className="min-height-full content has-background-dark p-6 has-text-light">
			<h1 className="has-text-light">Todo App</h1>
			<hr className="full-rule has-background-grey-dark" />

			<h2 className="has-text-primary">Add task:</h2>

			<TodoAdd
				handleAddTodo={ handleAddTodo }
				/>

			<hr className="full-rule has-background-grey-dark" />
			<h2 className="has-text-primary">
				Tasks: <span className="has-text-primary-light">{ todos.length }</span>
			</h2>

			<TodoList
				todos={ todos }
				handleToggle={ handleToggle }
				handleDelete={ handleDelete }
				/>
		</div>
	);
};

export default TodoApp;
