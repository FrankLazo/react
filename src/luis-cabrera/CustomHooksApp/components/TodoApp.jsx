import useTodos from '../hooks/useTodos';

const initialTodos = [
  { id: 1, title: 'Learn React' },
  { id: 2, title: 'Learn Node.js' },
];

const TodoApp = () => {
  const [todos, addTodo, deleteTodo] = useTodos(initialTodos);

  return (
		<div className="block is-flex is-flex-direction-column is-align-items-center">
			<h2>Todo App</h2>

			<button
				className="button is-primary"
				onClick={ () => addTodo({ title: 'New To-Do' }) }>
				Add
			</button>

			<ul>
			{
				todos.map(todo => (
					<li
						key={ todo.id }
						className="is-flex is-align-items-center">
						{ todo.title }
						<button
							className="button is-primary is-outlined ml-4"
							onClick={ () => deleteTodo(todo.id) }>
							Delete
						</button>
					</li>
				))
			}
			</ul>
		</div>
	);
}

export default TodoApp;
