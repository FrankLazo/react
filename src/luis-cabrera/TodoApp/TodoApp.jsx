import { useEffect, useState } from "react";
import TodoForm from "./components/TodoForm";
import TodoList from "./components/TodoList";

const initialTodos = [
	{
		id: 1,
		title: 'Todo #1',
		description: 'Descripci贸n del todo #1',
		completed: true,
	},
	{
		id: 2,
		title: 'Todo #2',
		description: 'Descripci贸n del todo #2',
		completed: true,
	},
	{
		id: 3,
		title: 'Todo #3',
		description: 'Descripci贸n del todo #3',
		completed: false,
	},
	{
		id: 4,
		title: 'Todo #4',
		description: 'Descripci贸n del todo #4',
		completed: false,
	},
];

const localTodos = JSON.parse(localStorage.getItem('todos'));

const TodoApp = () => {
	const [todos, setTodos] = useState(localTodos || initialTodos);
	const [todoEdit, setTodoEdit] = useState(null);

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	const todoDelete = (todoId) => {
		if (todoEdit && todoEdit.id === todoId)
		{
			setTodoEdit(null);
		}

		const changedTodos = todos.filter(todo => (todo.id !== todoId));
		setTodos(changedTodos);
	};

	const todoToggleCompleted = (todoId) => {
		const changedTodos = todos.map(todo => (
			todo.id === todoId
			? {...todo, completed: !todo.completed}
			: todo
		));

		setTodos(changedTodos);
	};

	const todoAdd = (todo) => {
		const newTodo = {
			id: Date.now(),
			...todo,
			completed: false,
		};

		const changedTodos = [
			...todos,
			newTodo,
		];

		setTodos(changedTodos);
	};

	const todoUpdate = (todoEdit) => {
		const changedTodos = todos.map(todo => (
			todo.id === todoEdit.id
			? todoEdit
			: todo
		));

		setTodos(changedTodos);
	}

	return (
		<div className="todoApp container content has-background-light p-4">
			<h1 className="has-text-primary">Todo app</h1>

			<div className="columns">
				<TodoList
					todos={todos}
					setTodoEdit={setTodoEdit}
					todoDelete={todoDelete}
					todoToggleCompleted={todoToggleCompleted}
					/>

				<TodoForm
					todoEdit={todoEdit}
					setTodoEdit={setTodoEdit}
					todoAdd={todoAdd}
					todoUpdate={todoUpdate}
					/>
			</div>
		</div>
	)
}

export default TodoApp;
