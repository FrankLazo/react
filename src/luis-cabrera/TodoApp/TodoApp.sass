.todoApp
	min-height: 100vh
