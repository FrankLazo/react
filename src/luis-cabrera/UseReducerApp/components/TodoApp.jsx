import { useReducer, useState } from "react";

const types =
{
	add: 	'add',
	update:	'update',
	delete: 'delete',
}

const initialTodos =
[
	{ id: 1, title: 'Todo #1', },
	{ id: 2, title: 'Todo #2', },
	{ id: 3, title: 'Todo #3', },
]

const reducer = (state, action) =>
{
	switch (action.type)
	{
		case types.delete:
			return state.filter(todo => todo.id !== action.payload);

		case types.add:
			return [...state, action.payload];

		case types.update:
		{
			const todoEdit = action.payload;
			return state.map(todo => todo.id === todoEdit.id ? todoEdit : todo);
		}

		default:
			return state;
	}
}

const TodoApp = () =>
{
	const [todos, dispatch] = useReducer(reducer, initialTodos);
	const [text, setText] = useState('');

	const handleSubmit = (e) =>
	{
		e.preventDefault();

		const newTodo =
		{
			id: Date.now(),
			title: text,
		}

		dispatch({ type: types.add, payload: newTodo });
	}

	return (
		<div className="block is-flex is-flex-direction-column is-align-items-center">
			<h2>Todo App</h2>

			<ul>
			{
				todos.map(todo => (
					<li key={ todo.id } className="is-bulletless is-flex is-align-items-center">
						{ todo.title }

						<div className="field has-addons ml-4">
							<div className="control">
								<button
									onClick={ () => dispatch({
										type: types.delete,
										payload: todo.id,
									})}
									className="button is-primary is-light"
									>
									Delete
								</button>
							</div>

							<div className="control">
								<button
									onClick={ () => dispatch({
										type: types.update,
										payload: { ...todo, title: text },
									})}
									className="button is-primary"
									>
									Update
								</button>
							</div>
						</div>
					</li>
				))
			}
			</ul>

			<form className="form"
				onSubmit={ handleSubmit }
				>
				<div className="field">
					<div className="control">
						<input
							type="text"
							value={ text }
							placeholder="Todo"
							onChange={ e => setText(e.target.value) }
							className="input"
							/>
					</div>
				</div>
			</form>
		</div>
	);
}

export default TodoApp;
