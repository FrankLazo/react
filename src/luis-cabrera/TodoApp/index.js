import TodoApp from "./TodoApp";

const index = () => {
	return (
		<TodoApp />
	);
}

export default index;
