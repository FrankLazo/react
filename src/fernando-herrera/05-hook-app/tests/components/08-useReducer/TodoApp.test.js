import { act } from '@testing-library/react';
import { shallow, mount } from 'enzyme';
import TodoApp from "../../../components/08-useReducer/TodoApp"
import demoTodos from '../../fixtures/demoTodos';

describe('<TodoApp /> tests', () => {
	const wrapper = shallow(<TodoApp />);

	Storage.prototype.setItem = jest.fn(() => {});

	test('should show the right component', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should add a todo', () => {
		const wrapper = mount(<TodoApp />);

		act(() => {
			wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] );
			wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[1] );
		});

		expect(wrapper.find('span').text().trim()).toBe('2');
		expect(localStorage.setItem).toHaveBeenCalledTimes(2);
	});

	test('should delete one todo', () => {
		wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] );
		wrapper.find('TodoList').prop('handleDelete')( demoTodos[0].id );

		expect(wrapper.find('span').text().trim()).toBe('0');
	});
});
