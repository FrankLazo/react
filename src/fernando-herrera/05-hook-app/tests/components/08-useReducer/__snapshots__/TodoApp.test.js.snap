// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoApp /> tests should show the right component 1`] = `
<div
  className="min-height-full content has-background-dark p-6 has-text-light"
>
  <h1
    className="has-text-light"
  >
    Todo App
  </h1>
  <hr
    className="full-rule has-background-grey-dark"
  />
  <h2
    className="has-text-primary"
  >
    Add task:
  </h2>
  <TodoAdd
    handleAddTodo={[Function]}
  />
  <hr
    className="full-rule has-background-grey-dark"
  />
  <h2
    className="has-text-primary"
  >
    Tasks: 
    <span
      className="has-text-primary-light"
    >
      0
    </span>
  </h2>
  <TodoList
    handleDelete={[Function]}
    handleToggle={[Function]}
    todos={Array []}
  />
</div>
`;
